### PR TITLE
Add CsrBuilder and CooBuilder for intrusive access to matrix arrays

### LINKED ISSUE #385: Intrusive access to matrix formats
Currently, I see no way how to modify the internal data structures of our matrix formats. By this I don't mean their values, but e.g. resizing the value array, as required for the SpGeMM and other methods modifying the sparsity pattern/nnz. As far as I can see, this is currently not possible from the public interface of the sparse formats?
How would you extend the interface? Aside from resizing, the current pointer-interface should handle all use cases, so adding a combined `resize_...` (or `set_nnz`?) method for the internal arrays that can change independent of the dimensions seems like a simple solution. Other ideas?

### PR DESCRIPTION
**Pull request by upsj**
These helper classes allow kernels to resize and otherwise manipulate the _variable-sized_ internal arrays of the `Csr` and `Coo` matrix types. Both operations are necessary for the ParILUT implementation.

TODO:

- [x] use them to simplify the `spgemm` interface

Closes #385

### CODE REVIEW

File: core/test/matrix/csr_builder.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        virtual int64_t clac_size(const int64_t nnz) { return 0; }
+
+        mock_strategy(bool &flag) : Mtx::strategy_type(""), was_called(&flag) {}
+
+        bool *was_called;
+    };
+    bool was_called{};
+    mtx->set_strategy(std::make_shared<mock_strategy>(was_called));
+
+    gko::matrix::CsrBuilder<>{mtx.get()};
```
Comment (tcojean) [inline core/test/matrix/csr_builder.cpp]: Is `was_called` supposed to be set on the destruction of this object? In that case, why does it work currently, shouldn't this be scoped inside a `{}` block? Currently, the destruction should happen at the very end of this block, after the `ASSERT_TRUE`.

Comment (tcojean) [reply core/test/matrix/csr_builder.cpp]: I think what you're testing is not destruction related, but the call to `make_srow` in `set_strategy` (which calls process further down).

Comment (upsj) [reply core/test/matrix/csr_builder.cpp]: > Is `was_called` supposed to be set on the destruction of this object? In that case, why does it work currently, shouldn't this be scoped inside a `{}` block? Currently, the destruction should happen at the very end of this block, after the `ASSERT_TRUE`.

The CsrBuilder is a temporary, as it doesn't have a variable name. Thus it get destroyed immediately.


> I think what you're testing is not destruction related, but the call to `make_srow` in `set_strategy` (which calls process further down). Maybe I misunderstood the test name.

That's correct, I added a fix.

Comment (tcojean) [reply core/test/matrix/csr_builder.cpp]: Ah right, I missed that this was an unnamed variable, thanks for clearing this up.

File: core/matrix/coo_builder.hpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * Allows intrusive access to the arrays stored within a @ref Coo matrix.
+ *
+ * @tparam ValueType  the value type of the matrix
+ * @tparam IndexType  the index type of the matrix
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class CooBuilder {
+public:
+    /** Returns the row index array of the COO matrix. */
```
Comment (pratikvn) [inline core/matrix/coo_builder.hpp]: This seems to formatted differently than all other member functions in other classes ?

Comment (upsj) [reply core/matrix/coo_builder.hpp]: Can you elaborate? This looks similar to other simple member functions like [Array::get_num_elems](https://github.com/ginkgo-project/ginkgo/blob/cbfae96ab838af1a5e020557025c990c0e451e50/include/ginkgo/core/base/array.hpp#L388)

**Comment by yhmtsai** [comment]
Is there any consideration about adding these function into new class `CooBuilder` not original class `Coo`?

**Comment by upsj** [comment]
The main argument is that we don't want to expose any `Array` to our users. Instead, we have these helpers that should only be used directly by kernels (which is also why I placed them in `core` and not `include/ginkgo/core`)

File: core/matrix/coo_builder.hpp (hunk 2 above)
Comment (pratikvn) [reply core/matrix/coo_builder.hpp]: I meant can it be like this instead:

```suggestion
    /**
     * Returns the row index array of the COO matrix.
     */
```

**Comment by yhmtsai** [comment]
It makes sense. Thanks

File: core/matrix/coo_builder.hpp (hunk 2 above)
Comment (upsj) [reply core/matrix/coo_builder.hpp]: I see, that makes sense. I changed the comments.

**Pull request merged**